### DISCUSSION TITLE
post_232603

### DISCUSSION THREAD

**Discussion by cincuranet** [cincuranet/blog-comments#1168] [Comments]
imported 

https://www.tabsoverspaces.com/id/232603

**Comment by cincuranet** [comment]
Imported comment written by _André_ on _2012-09-24T14:07:47_:

<p>Can we have a more deeper logging? I'm using this but it doesn't really show me the required details:</p><p>I get the error message "The type of the key field 'ID' is expected to be 'System.Int32', but the value provided is actually of type 'System.Int64'." and would like to see more details (complete callstack with all parameters). Can this be done? I have no idea why this happens because everything is int (in DB, in EDMX model, in generated EF classes)</p><p>Here is what I currently get from logging:</p><p>FirebirdSql.Data.FirebirdClient Information: 0 : Command:<br>SELECT FIRST (2) <br>"B"."ID" AS "ID",</p><p>*removed the noise*</p><p>FROM "User" AS "B"<br>WHERE "B"."ID" = @p__linq__0<br>FirebirdSql.Data.FirebirdClient Information: 0 : Parameters:</p><p>FirebirdSql.Data.FirebirdClient Information: 0 : Name:p__linq__0Type:Integer</p>

**Comment by cincuranet** [comment]
Imported comment written by _Jiří Činčura ↹_ on _2012-09-24T15:39:26_:

<p>This is error from Entity Framework. I think command logging will not help you here. But of course you can grab sources and add whatever you want.</p>

**Comment by cincuranet** [comment]
Imported comment written by _André_ on _2012-10-02T16:37:36_:

<p>the DB was corrupted. I've restored an older backup and I no longer get error messages.</p><p>But I have a new issue. When I create an empty EF model and try to generate the model from a Firebird DB, the associations are not retrieved correctly. Now all entries in all tables have the primary key symbol in the model designer and I get messages that:</p><p>Message2The relationship '' has columns that are not part of the key of the table on the primary side of the relationship. The relationship was excluded.D:\ConsoleApplication1\ConsoleApplication1\Model.edmx00ConsoleApplication1</p><p>Can I activate logging for VS2010? Can I add the entries to the devenv.exe.config so that the SQL commands are logged?</p><p>I've tried the latest Firebird 2.5.3 snapshot (2.5.3.26543) and got the same result.</p>

**Comment by cincuranet** [comment]
Imported comment written by _Jiří Činčura ↹_ on _2012-10-02T16:40:15_:

<p>You can see these in Output Window.</p>

**Comment by cincuranet** [comment]
Imported comment written by _André_ on _2012-10-02T16:53:33_:

<p>this is the only output I currently see:</p><p>The model was generated with warnings or errors.<br>Please see the Error List for more details. These issues must be fixed before running your application.<br>Loading metadata from the database took 00:00:05.6906883.<br>Generating the model took 00:00:01.7163335.<br>Added the connection string to the App.Config file.</p>

**Comment by cincuranet** [comment]
Imported comment written by _André_ on _2012-10-02T17:37:58_:

<p>ok, I've added the logging configuration to the config file of SV2010. I got the SQL command and now I need to understand what all those columns mean (c1 - c11).</p><p>c1 = ?<br>c2 = name of db column<br>c3 = is NULL<br>c4 = data type<br>c5 = max length<br>c6 = ?<br>c7 = ?<br>c8 = ?<br>c9 = ID (here everything is 0, it doesn't matter if the column is assigned to be a PK or not)<br>c10 = generator based (here everything is 0, it doesn't matter if the column is assigned to be a PK + Generator based generation or not)<br>c11 = ? (here everything is 1)</p>

**Comment by cincuranet** [comment]
Imported comment written by _André_ on _2012-10-02T18:12:01_:

<p>Here I can see that other users have the same issue:</p><p><a href="http://tracker.firebirdsql.org/browse/DNET-438" rel="nofollow noopener" title="http://tracker.firebirdsql.org/browse/DNET-438">http://tracker.firebirdsql....</a></p>

**Comment by cincuranet** [comment]
Imported comment written by _André_ on _2012-10-03T14:22:57_:

<p>I've tried the 2.5.1 ad 2.5.2 RC again and today it works, but the 2.5.3 snapshots are all broken. I've tried 2.5.3.26540 and 2.5.3.26543 and all generate the issue. But this is wired. Yesterday it didn't worked with the RC. And the user had the issue in June and he probably used the 2.5.1 final.</p><p>Btw, can the logging also include the returned values?</p>

**Comment by cincuranet** [comment]
Imported comment written by _André_ on _2012-12-07T15:32:23_:

<p>Can this be used to get the generated SQL commands before calling SaveChanges in EF (DbContext)?</p>

**Comment by cincuranet** [comment]
Imported comment written by _Jiří Činčura ↹_ on _2012-12-07T17:02:54_:

<p>Yes, it can.</p>

**Comment by cincuranet** [comment]
Imported comment written by _André_ on _2012-12-07T20:29:00_:

<p>for me it doesn't work. I can't see the SQL Statement that I'm interested in. I can see other statements, that I don't care. I get an error when calling SaceChanges that I want to debug by looking at the SQL Code:</p><p>"The operation failed: The relationship could not be changed because one or more of the foreign-key properties is non-nullable. When a change is made to a relationship, the related foreign-key property is set to a null value. If the foreign-key does not support null values, a new relationship must be defined, the foreign-key property must be assigned another non-null value, or the unrelated object must be deleted."</p><p>So I want to stop the EF work at SaveChanges and look at the generated SQL statement instead to see which values may cause the issue. I found this, but it is for SQL Server: <a href="http://www.codeproject.com/Articles/499902/Profiling-Entity-Framework-5-in-code" rel="nofollow noopener" title="http://www.codeproject.com/Articles/499902/Profiling-Entity-Framework-5-in-code">http://www.codeproject.com/...</a></p>

**Comment by cincuranet** [comment]
Imported comment written by _Jiří Činčura ↹_ on _2012-12-07T21:47:21_:

<p>This is error from EF even before sending data to database. Thus no command is executed.</p>

**Comment by cincuranet** [comment]
Imported comment written by _André_ on _2012-12-08T14:35:13_:

<p>I know. That's why I want to know if it is possible to skip EF validations and let EF generate the SQL statement and print it out.</p>

**Comment by cincuranet** [comment]
Imported comment written by _Jiří Činčura ↹_ on _2012-12-08T16:07:56_:

<p>No. If the model is not valid, it's imposible to generate statements.</p>

**Comment by cincuranet** [comment]
Imported comment written by _Christian A Pradelli_ on _2019-04-15T18:30:08_:

<p>Where should I set this configuration when I use a dotnet core 2.2 console application?<br>I try to create an app.config file and put that inside but it doesn't work.</p><p>Regards</p>

**Comment by cincuranet** [comment]
Imported comment written by _Jiří Činčura ↹_ on _2019-05-07T15:53:43_:

<p>Not sure. I think you have to do it in code (only). (Sorry for late reply, the reply somewhat didn't get through Disqus.)</p>

**Comment by krilbe** [comment]
Any reason this wouldn't work with Nuget package 7.1.1 and VS 207?
I know it has been working before, but as it is now, my Output panel doesn't seem to receive any SQL traces whatsoever.

**Comment by krilbe** [comment]
I meant VS 2017. Thought I'd try with 8.5.4 but it seems to bring requirements that cause problems in my project that I don't have time to resolve right now.

**Comment by cincuranet** [comment]
You want to check https://github.com/FirebirdSQL/NETProvider/issues/721.

**Comment by krilbe** [comment]
OK, so what does that mean? How do I enable it?
I tried changing to `<source name="FirebirdSql.Data.Logging">....`  but I still can't see any trace output.

**Comment by cincuranet** [comment]
It's no longer based on config (or trace sources), it's code based. Check the commit in the linked issue to see the changes.

**Comment by krilbe** [comment]
OK, I added this to Global.asax.cs in Application_Start:

`FirebirdSql.Data.Logging.FbLogManager.Provider = new FirebirdSql.Data.Logging.ConsoleLoggerProvider(logLevel);
FirebirdSql.Data.Logging.FbLogManager.IsParameterLoggingEnabled = true;`

The level is set to Debug. Nothing appears in the Output panel in VS2017. Is the above wrong? Do I need to do anything else?

I did test both with and without this in web.config:

`<system.diagnostics>
  <trace autoflush="false" indentsize="4">
    <listeners>
      <add name="configConsoleListener" type="System.Diagnostics.ConsoleTraceListener" />
    </listeners>
  </trace>
</system.diagnostics>`

**Comment by cincuranet** [comment]
`ConsoleLoggerProvider` logs, as name suggests, to console. If you want other target you need to provide your own implementation.
